Fix for dev dependencies overriding the always dependencies

gfusee opened the pull request:
## Description 

This PR intends to fix issue #23173. In order to address the issue, I chose the following strategy:

1) Add the utility function `BuildConfig::get_dependency_mode`.
2) Bring the mode's context into the `DependencyGraphBuilder::get_graph` function. This function is always called in places where a `BuildConfig` instance exists, making it easy to add this information.
3) If the mode is `DevOnly`, the logic doesn't change.
4) If the mode is `Always`, we don't compute the dev dependencies and address graphs, but provide empty ones.

This approach is conservative, as it doesn't change the logic but only the input data depending on the mode. Nevertheless, the PR is not 100% finished and I want to do the following:

- Add more unit tests to cover issue #23173
- Investigate the `get_graph -> collect_graphs -> get_graph` calls. From my understanding of your codebase, this is the only place where my PR could have a side effect. I haven't had the time yet to check it.
- Check if filtering the graph in `DependencyGraph::immediate_dependencies` is still useful. I also haven't had the time yet to check this.

## Test plan 

Manual tests, but automated tests are planned.

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol  
- [ ] Nodes (Validators and Full nodes)  
- [ ] gRPC  
- [ ] JSON-RPC  
- [ ] GraphQL  
- [x] CLI: fixes #23173  
- [ ] Rust SDK

gfusee commented:
@stefan-mysten great, I'm going to fix the other issues later today 🙏

hoangquocvietuet commented:
With this fix, the `Move.lock` file won’t stay consistent. I suggest including all packages in the dependencies and pairing them with the build mode as the key in the BTreeMap.

gfusee commented:
I fixed the issues you mentioned, @stefan-mysten. I also added a test called `always_dep_when_similar_dev_dep` and split the `immediate_dependencies` test into two tests: `immediate_dependencies_always` and `immediate_dependencies_dev`.

@hoangquocvietuet, you're right. A simpler suggestion for this would be, instead of pairing in the key type, to have a `Move.dev.lock` file dedicated to dev builds. This would make the developer experience less confusing and the Sui codebase simpler.

Or in a similar fashion, changing the Move.lock structure to add a `move.dev-package` array that would be read and written when the `--dev` flag is passed, instead of `move.package`.

hoangquocvietuet commented:
> This would make the developer experience less confusing and the Sui codebase simpler

I don’t quite agree with this. Using pairs is straightforward and makes sense. Adding more fields, such as an array, is also a good approach. 

However, introducing additional files would break a lot of things, including existing Sui Move courses that have already been published.

gfusee commented:
The advantage of a `move.dev-package` array in `Move.lock` is that the developer can review it intuitively. After sleeping on it, I agree with you that adding additional files, such as a `Move.dev.lock`, is not ideal.

@stefan-mysten What do you think about adding a move.dev-package to represent the pinned dependencies when `--dev` is passed? It would be similar in form to `move.package`.

stefan-mysten commented:
Thanks @gfusee - I kicked off the CI again.

stefan-mysten commented:
> adding a move.dev-package to represent the pinned dependencies when --dev is passed? It would be similar in form to move.package.

How would you deal with existing lock files where this does not exist? I am a bit torn, on one hand because we're working on a new pkg system which is close to being ready where there's no more concept of dev, lock files have a very different format, etc. So I feel that we can be ok without this up to the moment when we launch the new system.
On the other hand, might be good to think about it.

stefan-mysten commented:
@gfusee - I don't think this change works correctly. You can try to run `cargo nextest run` on `move-package` crate and you'll see a bunch of errors. I'll try to have a closer look to understand what's going on.

gfusee commented:
Indeed, the test-runner tests are failing. I have to check how those tests work, as they seem to read from external files to assert things.

For the dev_package field, we could do something similar to what is currently done in this PR. First, in `Packages::read`, we also return the dev packages in the tuple. Finally, we could change `DependencyGraph::read_from_lock` to specify the mode. If mode == `DevOnly`, we chain the iterator `packages.packages.into_iter().flatten()` with `dev_packages.packages.into_iter().flatten()`. This would make the dev pinned deps to override the always pinned deps.
